### PR TITLE
Contest Edit

### PR DESCRIPTION
**Pull request by twd2**

### CODE REVIEW

File: vj4/handler/contest.py [hunk 1]
```diff
@@ -337,3 +337,57 @@ class ContestCreateHandler(base.Handler, ContestStatusMixin):
...
+  async def post(self, *, tid: objectid.ObjectId, title: str, content: str, rule: int,
+                 begin_at_date: str,
+                 begin_at_time: str,
+                 duration: float,
+                 pids: str):
+    tdoc = await contest.get(self.domain_id, tid)
+    if not self.own(tdoc, builtin.PERM_EDIT_CONTEST_SELF):
+      self.check_perm(builtin.PERM_EDIT_CONTEST)
+    try:
+      begin_at = datetime.datetime.strptime(begin_at_date + ' ' + begin_at_time, '%Y-%m-%d %H:%M')
```
Comment (breeswish) [inline vj4/handler/contest.py]: Is it possible to reuse the logic here? i.e. can we introduce a new type hint for data times?

File: vj4/handler/contest.py [hunk 2]
```diff
@@ -337,3 +337,57 @@ class ContestCreateHandler(base.Handler, ContestStatusMixin):
...
+      begin_at = self.timezone.localize(begin_at).astimezone(pytz.utc).replace(tzinfo=None)
+      end_at = begin_at + datetime.timedelta(hours=duration)
+    except ValueError as e:
+      raise error.ValidationError('begin_at_date', 'begin_at_time')
+    if begin_at >= end_at:
+      raise error.ValidationError('duration')
```
Comment (breeswish) [inline vj4/handler/contest.py]: This seems to be already validated in model/adaptor/contest.

Comment (twd2) [reply vj4/handler/contest.py]: Model does not know anything about `duration` but we need to raise ValidationError for `duration` here.

File: vj4/handler/contest.py (hunk 1 above)
Comment (twd2) [reply vj4/handler/contest.py]: swx template :) :(

**Comment by twd2** [comment]
@breeswish merge me

**Pull request merged**